.github/workflows/phpunit-tests.yml

```diff
@@ -124,31 +121,25 @@ jobs:
 
       - name: Get composer cache directory
         id: composer-cache
-        if: ${{ env.COMPOSER_INSTALL == true || env.LOCAL_PHP == '8.0-fpm' }}
         run: echo "::set-output name=dir::$(composer config cache-files-dir)"
 
       - name: Cache Composer dependencies
-        if: ${{ env.COMPOSER_INSTALL == true || env.LOCAL_PHP == '8.0-fpm' }}
         uses: actions/cache@26968a09c0ea4f3e233fdddbafd1166051a095f6 # v2.1.4
         env:
           cache-name: cache-composer-dependencies
         with:
           path: ${{ steps.composer-cache.outputs.dir }}
-          key: ${{ runner.os }}-php-${{ matrix.php }}-composer-${{ hashFiles('**/composer.lock') }}
+          key: ${{ runner.os }}-php-${{ matrix.php }}-composer-${{ hashFiles('**/composer.json') }}
```

Comment:
I think using `composer.json` will be our only option here, but this could result in the the latest dependency versions within the defined ranges being used for some time. The cache key would either need to be evicted, or the file updated.

Probably not a huge deal. It would be nice if GH had a way to manually flush a cache. in order to ensure new versions of dependencies are used.

Reply:
> this could result in the the latest dependency versions within the defined ranges being used for some time
> ...
> in order to ensure new versions of dependencies are used.

I'm not terribly familiar with how the caching works in GHA, so bear with while we hash this out...

If I understand it correctly:
* Composer caching should cache the directory to which Composer _downloads_ the package zips files.
* Without a `lock` file, Composer will effectively always run `composer update`, even when `composer install` would be called.
* In the Composer process flow, it will:
    - first resolve the version constraints and available package version to the versions it should install;
    - then look in the cache if a download of the package(s) at the desired version is already available and install from that;
    - if not, it will download the version of the package which it needs and install it. The newly downloaded version is added to the downloads directory.
* Any updates to the `downloads` directory should update the cache file.

If this is correct and the downloads directory is the directory being cached and the cache is updated in GHA when that directory has changed:
* We should always have up to date dependencies installed within the constraints set in the `composer.json` file.
* The cache will grow until at some point it is invalidated.
* A change in the `composer.json` file will invalidate the cache and create a clean cache.

What am I missing ?

Reply:
> Any updates to the downloads directory should update the cache file

This is how caching works on Travis but unfortunately not on GHA. An existing cache cannot be updated, only a new cache can be created by altering the cache key.

From [the docs here](https://docs.github.com/en/actions/guides/caching-dependencies-to-speed-up-workflows#example-using-the-cache-action):

> Once you create a cache, you cannot change the contents of an existing cache but you can create a new cache with a new key.

Reply:
Thanks @johnbillion for the clarification.

In that case and combining that information with what I outlined above: over time CI builds may become marginally slower as Composer will download more files for newer versions of packages which have become available and are not in the cache, but the actual runs would still use the most up-to-date dependencies. Correct ?

If we then look at the current dependencies and the version constraints I've set in the `composer.json` file and to summarize what is largely already listed above in the commit information for the "remove `composer.lock`" commit:
* The `dealerdirect/phpcodesniffer-composer-installer` dependency doesn't release very often as it just works ;-)
    The current version constraints will only allow for a next `0.7.x` version anyhow.
* The `squizlabs/php_codesniffer` dependency releases every ~3 months or so, but the version is fixed, so this would not impact the caching as an update would require a change to the `composer.json` file.
* The `wp-coding-standards/wpcs` dependency is waiting on the next major, after which it is likely to be fixed to a set version, making it a managed dependency which would need an update to the `composer.json` file.
* The `phpcompatibility/phpcompatibility-wp` dependency only updates when either WP adds a polyfill for a PHP feature or when there is a major release of the underlying `PHPCompatibility` package. Again, not really an issue.
* The underlying `PHPCompatibility` dependency is waiting on the next major. Releases can be relatively frequent in the six months around the release of a new PHP version.
    This is a likely dependency to cause the cache to be outdated (after the next major has been released).
* The `phpunit/phpunit` dependency and its underlying dependencies are not something I'm too concerned about at this time as the currently allowed versions are no longer supported and won't ever receive updates.
    Some of the underlying dependencies _may_ still receive updates, but that will be a small subset which still receive updates for the versions set in the constraints for PHPUnit 5.x and 7.x.
    Once the PHPUnit version constraints have been widened, it will be a different matter, but fixing all the underlying dependencies of PHPUnit in the `composer.json` file is unrealistic and undesirable.
    
If the CI cache going out of date is a serious concern, we need to have a good think about two options:
1. Possibly adding a timestamp into the cache key - something like "monday of this week at 12:00 UTC" or "first monday of the month at 12:00 UTC".
    That would evict the cache once every week/every month.
2. Instead of removing the `composer.lock` file, having separate `composer.lock` files for the various supported PHP versions. This is not something I can recommend as chances of something going wrong there with updates as well as the maintenance burden, is not something I'd like to contemplate. 

Opinions ?

Reply:
I've implemented option one now by adding "last Monday"'s date to the cache key, which means that the cache should be refreshed at least once a week (more often if the `composer.json` file is changed during the week).

@desrosj @johnbillion What do you think ?

Reply:
I'm wondering whether we shouldn't just switch to using the [Composer caching action](https://github.com/ramsey/composer-install) for this.
I get the impression that that action includes a hash of all the installed versions in the cache key, which would make it evict every time a new dependency becomes available, which is basically what we want.

It would also simplify/clean up the script as we can remove the all the steps involved in caching the dependencies now.

I've opened a question in that repo for further clarification: https://github.com/ramsey/composer-install/issues/164

Reply:
The "last Monday" date is clever. I like it!

ramsey/composer-install is also interesting. But I think we could move forward and switch to that at a later date should it seem like it better suits us.

Reply:
I've squash the CI commits and given it a proper commit message now (and updated the PR description with the commit message).

Reply:
For the record - I've opened Trac ticket https://core.trac.wordpress.org/ticket/53841 to investigate switching to the `ramsey/composer-install` action.